modules/schema/src/main/java/org/apache/ignite/internal/schema/mapping/ColumnMapper.java

```diff
@@ -24,6 +24,7 @@
 /**
  * Column mapper interface.
  */
+// TODO: IGNITE-15480 Probably should be removed after index column mapping refactoring.
```

Comment:
What do you mean?
The ticket has empty description.

Reply:
I mean that index should be mapped to columns by column_id's that will be a part of internal configuration extension. column_id's will be final, so column renaming won't effect column-to-index mapping.

Reply:
Internal column sort order differs from the user provided in config.
Now, column index is an index in array of columns.

So, we need mapping name->columnID->columnIndex. There is a ticket for this [1].

[1] https://issues.apache.org/jira/browse/IGNITE-13667

Reply:
I believe that within column to column in index context we don't need columnIndex (if you mean that column index here is position in columns list).

From the user's perspective he creates index using column names: 
```
        SchemaTable scmTbl1 = SchemaBuilders.tableBuilder("PUBLIC", "tbl1").columns(
            SchemaBuilders.column("key", ColumnType.UUID).asNonNull().build(),
            SchemaBuilders.column("affKey", ColumnType.INT64).asNonNull().build(),
            SchemaBuilders.column("valStr", ColumnType.string()).asNullable().build(),
            SchemaBuilders.column("valInt", ColumnType.INT32).asNullable().build(),
            SchemaBuilders.column("valNull", ColumnType.INT16).asNullable().build()
        ).withIndex(
            SchemaBuilders.pkIndex()
                .addIndexColumn("key").done()
                .addIndexColumn("affKey").done()
                .withAffinityColumns("affKey")
                .build()
        ).build();
```
However during configuration change we generate id for every column and substitute name with such id within "index schema":
table
  name:PUBLIC.tbl1
  id: 41f7a3ca
  columns
    listItem0
      namedListKey: 0 (or "key" or 870024f6) it doesn't matter here 
      namedListVal:
        name: key
        **id: 870024f6**
        type: UUID
        nullable: false
   ...
   indexes
     ....
       name: PK_INDEX
       **columnIds: [870024f6]** instead of colNames

Does it make sence?

Reply:
Assume there is a row R with schema V1.
Then you update schema to V2 and add a column C with a default D1.
Then you update schema to V3 and update default to D2.

Now, you read R, which is stored as of V1 version, and upgrade it to V3. 
On one side, R doesn't contains C column, and there is no descriptor for column C in V1.
On another side, you can't use column C descriptor from V3 because it was changed in between.
Using column C descriptor from V2 will be correct, but you will need to go through schema history for that for every row and every column you read just to check if there is any history.
Column mapper merges schema changes once and then get reused.

Reply:
You suggestion about columnID look great, it makes add/drop simple, resolves ABA problem and gives rename for free.
But we will need mapping columnID -> columnIdx in schemaDescriptor internals anyway.

Reply:
> we don't need columnIndex (if you mean that column index here is position in columns list).
I mean column index in Columns class. E,g, see SchemaDescriptor.keyCols.
We need a specific order for serialization: key columns prior to value columns and fixlen types prior to varlen types.
Also, we want null-map to be a bitset.

Reply:
As was discussed mapping logic was restored:
```
                clusterCfgMgr.configurationRegistry()
                    .getConfiguration(TablesConfiguration.KEY).tables()
                    .change(ch -> {
                        ch.createOrUpdate(name, tableChange);
                        ch.createOrUpdate(name, tblCh ->
                            ((ExtendedTableChange)tblCh).changeSchemas(
                                schemasCh ->
                                    schemasCh.createOrUpdate(
                                        String.valueOf(schemasCh.size() + 1),
                                        schemaCh -> {
                                            ExtendedTableView currTableView = (ExtendedTableView) clusterCfgMgr.configurationRegistry().getConfiguration(TablesConfiguration.KEY).tables().get(name).value();

                                            SchemaDescriptor descriptor = SchemaUtils.prepareSchemaDescriptor(
                                                ((ExtendedTableView)tblCh).schemas().size(),
                                                tblCh
                                            );

                                            descriptor.columnMapping(SchemaUtils.columnMapper(
                                                tablesById.get(tblId).schemaRegistry().schema(currTableView.schemas().size() -1),
                                                currTableView,
                                                descriptor,
                                                tblCh
                                            ));

                                            schemaCh.changeSchema(ByteUtils.toBytes(descriptor));
                                        }
                                    )
                            ));
                    })
```